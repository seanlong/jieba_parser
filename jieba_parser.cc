/*
 * PotgreSQL FTS parser extension based on Jieba
 * Copyright (c) 2014, Long Xiang(seanlong@outlook.com)
 * Copyright (c) 2007-2011, PostgreSQL Global Development Group
 */

extern "C" {

#include "postgres.h"

#include "catalog/pg_type.h"
#include "fmgr.h"
#include "tsearch/ts_type.h"
#include "utils/array.h"
#include "utils/builtins.h"

}

#undef _XOPEN_SOURCE
#undef _POSIX_C_SOURCE
#include <Python.h>

#include <memory>
#include <string>
#include <vector>

extern "C" {

PG_MODULE_MAGIC;

typedef struct {
	int			lexid;
	char	   *alias;
	char	   *descr;
} LexDescr;

/*
 * prototypes
 */
PG_FUNCTION_INFO_V1(jiebaprs_start);
Datum		jiebaprs_start(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(jiebaprs_getlexeme);
Datum		jiebaprs_getlexeme(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(jiebaprs_end);
Datum		jiebaprs_end(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(jiebaprs_lextype);
Datum		jiebaprs_lextype(PG_FUNCTION_ARGS);


PG_FUNCTION_INFO_V1(jieba_rank);
Datum   jieba_rank(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(jieba_extract_tags);
Datum jieba_extract_tags(PG_FUNCTION_ARGS);

}

namespace {

PyObject* ImportJiebaModule() {
  Py_Initialize();

  PyObject* name = PyString_FromString("jieba");
  PyObject* module = PyImport_Import(name);
  assert(module);
  Py_DECREF(name);

  char method[] = "initialize";
  PyObject* ret = PyObject_CallMethod(module, method, NULL);
  assert(ret);
  Py_DECREF(ret);

  return module;
}

struct ParserState {
  ParserState(PyObject* token_generator)
      : token_gen_(token_generator) {
  }

  ~ParserState() {
    ClearTokens();
    Py_DECREF(token_gen_);
  }

  char* NextToken() {
    PyObject* item = PyIter_Next(token_gen_);
    if (!item)
      return NULL;
    PyObject* str = PyUnicode_AsUTF8String(item);
    Py_DECREF(item);
    token_strs_.push_back(str);
    return PyString_AsString(str);
  }

  void ClearTokens() {
    auto iter = token_strs_.begin();
    for (; iter != token_strs_.end(); ++iter)
      Py_DECREF(*iter);
    token_strs_.clear();
  }

  PyObject* token_gen_;
  std::vector<PyObject*> token_strs_;
};

class JiebaParser {
 public:
  JiebaParser()
      : module_(NULL),
        cutter_(NULL) {
    PyObject* module = ImportJiebaModule();
    cutter_ = PyObject_GetAttrString(module, "cut_for_search");
    assert(cutter_);
    Py_DECREF(module);
  }

  ~JiebaParser() {
    Py_DECREF(cutter_);
    Py_DECREF(module_);
    Py_Finalize();
  }

  ParserState* Cut(const char* buffer, int length) {
    PyObject* args = PyTuple_New(1);
    PyTuple_SetItem(args, 0, PyUnicode_DecodeUTF8(buffer, length, NULL));
    ParserState* pst = new ParserState(PyObject_CallObject(cutter_, args));
    Py_DECREF(args);
    return pst;
  }

 private:
  PyObject* module_;
  PyObject* cutter_;
} jieba_parser;

class JiebaExtractor {
 public:
  JiebaExtractor()
      : tagger_(NULL),
        module_(NULL) {
    module_ = ImportJiebaModule();
    PyObject* analyse_name  = PyString_FromString("jieba.analyse");
    PyObject* analyse_module = PyImport_Import(analyse_name);
    assert(analyse_module);
    Py_DECREF(analyse_name);

    tagger_ = PyObject_GetAttrString(analyse_module, "extract_tags");
    assert(tagger_);
    Py_DECREF(analyse_module);
  }

  ~JiebaExtractor() {
    Py_DECREF(tagger_);
    Py_DECREF(module_);
    Py_Finalize();
  }

  std::unique_ptr<std::vector<std::string>>
  Tag(unsigned short top_num, const char* buffer, int length) {
    PyObject* kws = PyDict_New();
    PyDict_SetItemString(kws, "topK", PyInt_FromLong(top_num)); 
    PyObject* args = PyTuple_New(1);
    PyTuple_SetItem(args, 0, PyUnicode_DecodeUTF8(buffer, length, NULL));
    PyObject* tags = PyObject_Call(tagger_, args, kws);
    Py_DECREF(kws);
    Py_DECREF(args);

    assert(PyList_Check(tags));
    Py_ssize_t size = PyList_Size(tags);
    std::unique_ptr<std::vector<std::string>> ret(
        new std::vector<std::string>());
    for (int i = 0; i < size; ++i) {
      PyObject* item = PyList_GetItem(tags, i);  // borrowed reference
      PyObject* str = PyUnicode_AsUTF8String(item);  // new reference
      ret->push_back(PyString_AsString(str));  // internal buffer of string
      Py_DECREF(str);
    }
    return std::move(ret);
  }

 private:
  PyObject* tagger_;
  PyObject* module_;
} jieba_extractor_;

}

Datum jiebaprs_start(PG_FUNCTION_ARGS)
{
  char* buffer = PG_GETARG_POINTER(0);
  int len = PG_GETARG_INT32(1);
  PG_RETURN_POINTER(jieba_parser.Cut(buffer, len)); 
}

Datum jiebaprs_getlexeme(PG_FUNCTION_ARGS) {
  ParserState* pst = (ParserState*)PG_GETARG_POINTER(0);
	char** t = (char**)PG_GETARG_POINTER(1);
	int* tlen = (int*)PG_GETARG_POINTER(2);
  int type = 3;

  char* token = pst->NextToken();
  if (token) {
    *t = token;
    *tlen = strlen(token);
  } else {
    type = 0;
    *tlen = 0;
  }

	PG_RETURN_INT32(type);
}

Datum jiebaprs_end(PG_FUNCTION_ARGS) {
  ParserState* pst = (ParserState*)PG_GETARG_POINTER(0);
  delete pst;
  PG_RETURN_VOID();
}

Datum jiebaprs_lextype(PG_FUNCTION_ARGS) {
	LexDescr* descr = (LexDescr*)palloc(sizeof(LexDescr) * 2);
  descr[0].lexid = 3;
  descr[0].alias = pstrdup("word");
  descr[0].descr = pstrdup("Word");
  descr[1].lexid = 0;

	PG_RETURN_POINTER(descr);
}

Datum jieba_rank(PG_FUNCTION_ARGS) {
  TSVector txt = PG_GETARG_TSVECTOR(0);
  TSQuery query = PG_GETARG_TSQUERY(1);

  //TODO return rank based on td-idf.
  float res = 0.0;

  PG_FREE_IF_COPY(txt, 0);
  PG_FREE_IF_COPY(query, 1);
  PG_RETURN_FLOAT4(res);
}

Datum jieba_extract_tags(PG_FUNCTION_ARGS) {
  text* t = PG_GETARG_TEXT_P(0);
  uint16 top_num = PG_GETARG_UINT16(1);
  char* raw_text = text_to_cstring(t);
  auto tags = jieba_extractor_.Tag(top_num, raw_text, strlen(raw_text));

  Datum* tag_datums = (Datum*)palloc(tags->size() * sizeof(Datum));
  auto iter = tags->begin();
  for (int i = 0; iter != tags->end(); ++iter, ++i)
    tag_datums[i] = CStringGetTextDatum(iter->c_str());

  //fprintf(stderr, "%d %p %d\n", __LINE__, tag_datums, tags->size());
  ArrayType* array = construct_array(
      tag_datums, tags->size(), TEXTOID, -1, false, 'i');
  pfree(tag_datums);
  PG_RETURN_POINTER(array);
}
