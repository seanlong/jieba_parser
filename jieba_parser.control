# test_parser extension
comment = 'Jieba parser for full-text search'
default_version = '1.0'
module_pathname = '$libdir/jieba_parser'
relocatable = true
