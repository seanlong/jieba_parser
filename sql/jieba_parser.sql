CREATE EXTENSION jieba_parser;

-- make test configuration using parser

CREATE TEXT SEARCH CONFIGURATION jiebacfg (PARSER = jiebaparser);

ALTER TEXT SEARCH CONFIGURATION jiebacfg ADD MAPPING FOR word WITH simple;

-- ts_parse

SELECT * FROM ts_parse('jiebaparser', '小明硕士毕业于中国科学院计算所，后在日本京都大学深造');

SELECT to_tsvector('jiebacfg','我来到北京清华大学');

SELECT to_tsquery('jiebacfg', '清华');

SELECT to_tsvector('jiebacfg', '我来到北京清华大学') @@ to_tsquery('jiebacfg', '清华');

SELECT jieba_extract_tags('被称为“相声演员”的罗永浩，依然以调侃开场：先调侃了一下延迟开始所造成的“经济损失”，紧接着矛头直接指向眼下手机市场上的硬件大战，以及“跑分大战”；看上去，小米和雷军再次躺枪。', 5);
