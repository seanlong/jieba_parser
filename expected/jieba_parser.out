CREATE EXTENSION jieba_parser;
-- make test configuration using parser
CREATE TEXT SEARCH CONFIGURATION jiebacfg (PARSER = jiebaparser);
ALTER TEXT SEARCH CONFIGURATION jiebacfg ADD MAPPING FOR word WITH simple;
-- ts_parse
SELECT * FROM ts_parse('jiebaparser', '小明硕士毕业于中国科学院计算所，后在日本京都大学深造');
 tokid |    token     
-------+--------------
     3 | 小明
     3 | 硕士
     3 | 毕业
     3 | 于
     3 | 中国
     3 | 科学
     3 | 学院
     3 | 科学院
     3 | 中国科学院
     3 | 计算
     3 | 计算所
     3 | ，
     3 | 后
     3 | 在
     3 | 日本
     3 | 京都
     3 | 大学
     3 | 日本京都大学
     3 | 深造
(19 rows)

SELECT to_tsvector('jiebacfg','我来到北京清华大学');
                           to_tsvector                            
------------------------------------------------------------------
 '北京':3 '华大':5 '大学':6 '我':1 '来到':2 '清华':4 '清华大学':7
(1 row)

SELECT to_tsquery('jiebacfg', '清华');
 to_tsquery 
------------
 '清华'
(1 row)

SELECT to_tsvector('jiebacfg', '我来到北京清华大学') @@ to_tsquery('jiebacfg', '清华');
 ?column? 
----------
 t
(1 row)

SELECT jieba_extract_tags('被称为“相声演员”的罗永浩，依然以调侃开场：先调侃了一下延迟开始所造成的“经济损失”，紧接着矛头直接指向眼下手机市场上的硬件大战，以及“跑分大战”；看上去，小米和雷军再次躺枪。', 5);
      jieba_extract_tags      
------------------------------
 {调侃,大战,雷军,罗永浩,相声}
(1 row)

